In [16]:
#Regresion Lineal
#Arboles de decision
#random forest
import pandas as pd
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, f1_score
import pandas as pd
from sklearn.model_selection import train_test_split
import joblib

In [2]:
train_df = pd.read_csv('../Data/train.csv')
test_df = pd.read_csv('../Data/test.csv')
validation_df = pd.read_csv('../Data/validation.csv')

# Separate target column
train_X = train_df.drop(columns=['Depression'])
train_Y = train_df['Depression']

test_X = test_df.drop(columns=['Depression'])
test_Y = test_df['Depression']

validation_X = validation_df.drop(columns=['Depression'])
validation_Y = validation_df['Depression']

Se va a realizar un experimento donde se varian hiperparametros y se busca la mejor solucion

In [24]:
# Definir valores de los hiperparámetros para Random Forest
n_estimators_values = [50, 100, 200]
max_depth_values = [None, 10, 20, 30]
min_samples_split_values = [2, 5, 10]

# Lista para almacenar los resultados
results = []


Se entrena el modelo con todos los posibles valores de hiperparametros y se selecciona el que tenga mejores métricas


In [25]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, f1_score, confusion_matrix, precision_score, recall_score

# Loop para grid search manual
for n_estimators in n_estimators_values:
    for max_depth in max_depth_values:
        for min_samples_split in min_samples_split_values:
            # Inicializar el modelo Random Forest con los hiperparámetros
            model = RandomForestClassifier(n_estimators=n_estimators,
                                           max_depth=max_depth,
                                           min_samples_split=min_samples_split,
                                           random_state=42)

            # Entrenar el modelo
            model.fit(train_X, train_Y)

            # Hacer predicciones sobre el conjunto de validación
            validation_pred = model.predict(validation_X)

            # Calcular la matriz de confusión
            cm = confusion_matrix(validation_Y, validation_pred)

            # Calcular métricas
            accuracy = accuracy_score(validation_Y, validation_pred)
            precision = precision_score(validation_Y, validation_pred, average='weighted')
            recall = recall_score(validation_Y, validation_pred, average='weighted')
            f1 = f1_score(validation_Y, validation_pred, average='weighted')

            # Guardar los resultados en una lista
            results.append({
                'n_estimators': n_estimators,
                'max_depth': max_depth,
                'min_samples_split': min_samples_split,
                'Accuracy': accuracy,
                'Precision': precision,
                'Recall': recall,
                'F1 Score': f1,
                'Confusion Matrix': cm
            })

In [26]:
# Convert the results to a DataFrame
results_df = pd.DataFrame(results)
results_df

n_estimators  max_depth  min_samples_split  Accuracy  Precision    Recall  \
0             50        NaN                  2  0.849057   0.848315  0.849057   
1             50        NaN                  5  0.842767   0.842531  0.842767   
2             50        NaN                 10  0.842767   0.841976  0.842767   
3             50       10.0                  2  0.836478   0.835561  0.836478   
4             50       10.0                  5  0.861635   0.861518  0.861635   
5             50       10.0                 10  0.849057   0.848315  0.849057   
6             50       20.0                  2  0.849057   0.848648  0.849057   
7             50       20.0                  5  0.861635   0.861518  0.861635   
8             50       20.0                 10  0.842767   0.841976  0.842767   
9             50       30.0                  2  0.849057   0.848315  0.849057   
10            50       30.0                  5  0.842767   0.842531  0.842767   
11            50       30.0                 10  0.842767   0.841976  0.842767   
12           100        NaN                  2  0.842767   0.842393  0.842767   
13           100        NaN                  5  0.867925   0.868405  0.867925   
14           100        NaN                 10  0.861635   0.861068  0.861635   
15           100       10.0                  2  0.823899   0.822906  0.823899   
16           100       10.0                  5  0.823899   0.822906  0.823899   
17           100       10.0                 10  0.849057   0.848315  0.849057   
18           100       20.0                  2  0.849057   0.848315  0.849057   
19           100       20.0                  5  0.849057   0.848315  0.849057   
20           100       20.0                 10  0.867925   0.867635  0.867925   
21           100       30.0                  2  0.836478   0.835800  0.836478   
22           100       30.0                  5  0.867925   0.868405  0.867925   
23           100       30.0                 10  0.861635   0.861068  0.861635   
24           200        NaN                  2  0.836478   0.835777  0.836478   
25           200        NaN                  5  0.861635   0.862488  0.861635   
26           200        NaN                 10  0.842767   0.841976  0.842767   
27           200       10.0                  2  0.842767   0.842531  0.842767   
28           200       10.0                  5  0.842767   0.842531  0.842767   
29           200       10.0                 10  0.855346   0.855468  0.855346   
30           200       20.0                  2  0.836478   0.835777  0.836478   
31           200       20.0                  5  0.842767   0.842016  0.842767   
32           200       20.0                 10  0.842767   0.841976  0.842767   
33           200       30.0                  2  0.836478   0.835777  0.836478   
34           200       30.0                  5  0.861635   0.862488  0.861635   
35           200       30.0                 10  0.842767   0.841976  0.842767   

    F1 Score      Confusion Matrix  
0   0.848122  [[49, 14], [10, 86]]  
1   0.840908   [[47, 16], [9, 87]]  
2   0.842059  [[49, 14], [11, 85]]  
3   0.835466  [[48, 15], [11, 85]]  
4   0.860273   [[49, 14], [8, 88]]  
5   0.848122  [[49, 14], [10, 86]]  
6   0.847570   [[48, 15], [9, 87]]  
7   0.860273   [[49, 14], [8, 88]]  
8   0.842059  [[49, 14], [11, 85]]  
9   0.848122  [[49, 14], [10, 86]]  
10  0.840908   [[47, 16], [9, 87]]  
11  0.842059  [[49, 14], [11, 85]]  
12  0.842546  [[50, 13], [12, 84]]  
13  0.866363   [[49, 14], [7, 89]]  
14  0.860779   [[50, 13], [9, 87]]  
15  0.822165  [[46, 17], [11, 85]]  
16  0.822165  [[46, 17], [11, 85]]  
17  0.848122  [[49, 14], [10, 86]]  
18  0.848122  [[49, 14], [10, 86]]  
19  0.848122  [[49, 14], [10, 86]]  
20  0.866872   [[50, 13], [8, 88]]  
21  0.836002  [[49, 14], [12, 84]]  
22  0.866363   [[49, 14], [7, 89]]  
23  0.860779   [[50, 13], [9, 87]]  
24  0.834868  [[47, 16], [10, 86]]  
25  0.859712   [[48, 15], [7, 89]]  
26  0.8

In [30]:
# Select the best hyperparameter combination based on the R2 metric
best_params = results_df.loc[results_df['F1 Score'].idxmax()]
best_params

n_estimators                         100
max_depth                           20.0
min_samples_split                     10
Accuracy                        0.867925
Precision                       0.867635
Recall                          0.867925
F1 Score                        0.866872
Confusion Matrix     [[50, 13], [8, 88]]
Name: 20, dtype: object

Se entrena el modelo con los mejores parámetros.

In [31]:
# Train the model with the best hyperparameters on the validation set
best_model = RandomForestClassifier(n_estimators=int(best_params['n_estimators']),
                                   max_depth=int(best_params['max_depth']) if best_params['max_depth'] is not None else None,
                                   min_samples_split=int(best_params['min_samples_split']),
                                   random_state=42)

best_model.fit(validation_X, validation_Y)

RandomForestClassifier(max_depth=20, min_samples_split=10, random_state=42)

## Finalmente se usan los datos de "test" para validar las métricas del modelo.

In [32]:
# Evaluate the model on the test set
test_pred = best_model.predict(test_X)
test_mse = mean_squared_error(test_Y, test_pred)
test_r2 = r2_score(test_Y, test_pred)

# Display the final model metrics on the test set
print(f"Test MSE: {test_mse}")
print(f"Test R2: {test_r2}")

Test MSE: 0.17587939698492464
Test R2: 0.2549208386820708


Como ejemplo: se tiene un nuevo inmueble dado por los siguientes datos y se debe predecir el precio.

In [33]:
new_data = {
  "Age": [26.0],
  "Academic Pressure": [4.0],
  "CGPA": [8.78],
  "Study Satisfaction": [2.0],
  "Work/Study Hours": [5.0],
  "Financial Stress": [2.0],
  "Gender_Male": [0.0],
  "Dietary Habits_Moderate": [1.0],
  "Dietary Habits_Unhealthy": [0.0],
  "Degree_B.Arch": [0.0],
  "Degree_B.Com": [1.0],
  "Degree_B.Ed": [0.0],
  "Degree_B.Pharm": [0.0],
  "Degree_B.Tech": [0.0],
  "Degree_BA": [0.0],
  "Degree_BBA": [0.0],
  "Degree_BCA": [0.0],
  "Degree_BE": [0.0],
  "Degree_BHM": [0.0],
  "Degree_BSc": [0.0],
  "Degree_LLB": [0.0],
  "Degree_LLM": [0.0],
  "Degree_M.Com": [0.0],
  "Degree_M.Ed": [0.0],
  "Degree_M.Pharm": [0.0],
  "Degree_M.Tech": [0.0],
  "Degree_MA": [0.0],
  "Degree_MBA": [0.0],
  "Degree_MBBS": [0.0],
  "Degree_MCA": [0.0],
  "Degree_MD": [0.0],
  "Degree_ME": [0.0],
  "Degree_MSc": [0.0],
  "Degree_PhD": [0.0],
  "Have you ever had suicidal thoughts ?_Yes": [1.0],
  "Family History of Mental Illness_Yes": [0.0],
  "City_Ahmedabad": [0.0],
  "City_Bangalore": [0.0],
  "City_Bhopal": [0.0],
  "City_Chennai": [0.0],
  "City_Delhi": [0.0],
  "City_Faridabad": [0.0],
  "City_Ghaziabad": [0.0],
  "City_Hyderabad": [0.0],
  "City_Indore": [0.0],
  "City_Jaipur": [0.0],
  "City_Kalyan": [0.0],
  "City_Kanpur": [0.0],
  "City_Kolkata": [0.0],
  "City_Lucknow": [0.0],
  "City_Ludhiana": [0.0],
  "City_Meerut": [0.0],
  "City_Mumbai": [0.0],
  "City_Nagpur": [0.0],
  "City_Nashik": [0.0],
  "City_Patna": [0.0],
  "City_Pune": [0.0],
  "City_Rajkot": [0.0],
  "City_Srinagar": [0.0],
  "City_Surat": [0.0],
  "City_Thane": [0.0],
  "City_Vadodara": [0.0],
  "City_Varanasi": [0.0],
  "City_Vasai-Virar": [0.0],
  "City_Visakhapatnam": [0.0],
  "Sleep Duration_'7-8 hours'": [1.0],
  "Sleep Duration_'Less than 5 hours'": [0.0],
  "Sleep Duration_'More than 8 hours'": [0.0]
}



# Convertir el nuevo dato en un DataFrame
new_df = pd.DataFrame(new_data)

# Usar el modelo entrenado (best_model) para hacer la predicción
predicted_price = best_model.predict(new_df)

# Mostrar el resultado de la predicción
print(f"Predicted: {predicted_price[0]}")

Predicted: 1


In [34]:
new_data = {
  "Age": [33.0],
  "Academic Pressure": [5.0],
  "CGPA": [9.54],
  "Study Satisfaction": [1.0],
  "Work/Study Hours": [7.0],
  "Financial Stress": [2.0],
  "Gender_Male": [1.0],
  "Dietary Habits_Moderate": [0.0],
  "Dietary Habits_Unhealthy": [0.0],
  "Degree_B.Arch": [0.0],
  "Degree_B.Com": [0.0],
  "Degree_B.Ed": [0.0],
  "Degree_B.Pharm": [1.0],
  "Degree_B.Tech": [0.0],
  "Degree_BA": [0.0],
  "Degree_BBA": [0.0],
  "Degree_BCA": [0.0],
  "Degree_BE": [0.0],
  "Degree_BHM": [0.0],
  "Degree_BSc": [0.0],
  "Degree_LLB": [0.0],
  "Degree_LLM": [0.0],
  "Degree_M.Com": [0.0],
  "Degree_M.Ed": [0.0],
  "Degree_M.Pharm": [0.0],
  "Degree_M.Tech": [0.0],
  "Degree_MA": [0.0],
  "Degree_MBA": [0.0],
  "Degree_MBBS": [0.0],
  "Degree_MCA": [0.0],
  "Degree_MD": [0.0],
  "Degree_ME": [0.0],
  "Degree_MSc": [0.0],
  "Degree_PhD": [0.0],
  "Have you ever had suicidal thoughts ?_Yes": [1.0],
  "Family History of Mental Illness_Yes": [0.0],
  "City_Ahmedabad": [0.0],
  "City_Bangalore": [0.0],
  "City_Bhopal": [0.0],
  "City_Chennai": [0.0],
  "City_Delhi": [0.0],
  "City_Faridabad": [0.0],
  "City_Ghaziabad": [0.0],
  "City_Hyderabad": [0.0],
  "City_Indore": [0.0],
  "City_Jaipur": [0.0],
  "City_Kalyan": [0.0],
  "City_Kanpur": [0.0],
  "City_Kolkata": [0.0],
  "City_Lucknow": [0.0],
  "City_Ludhiana": [0.0],
  "City_Meerut": [0.0],
  "City_Mumbai": [0.0],
  "City_Nagpur": [0.0],
  "City_Nashik": [0.0],
  "City_Patna": [0.0],
  "City_Pune": [0.0],
  "City_Rajkot": [0.0],
  "City_Srinagar": [0.0],
  "City_Surat": [0.0],
  "City_Thane": [0.0],
  "City_Vadodara": [0.0],
  "City_Varanasi": [0.0],
  "City_Vasai-Virar": [0.0],
  "City_Visakhapatnam": [0.0],
  "Sleep Duration_'7-8 hours'": [0.0],
  "Sleep Duration_'Less than 5 hours'": [0.0],
  "Sleep Duration_'More than 8 hours'": [0.0]
}




# Convertir el nuevo dato en un DataFrame
new_df = pd.DataFrame(new_data)

# Usar el modelo entrenado (best_model) para hacer la predicción
predicted_price = best_model.predict(new_df)

# Mostrar el resultado de la predicción
print(f"Predicted: {predicted_price[0]}")

Predicted: 1


El modelo Random Forest parece tener un buen rendimiento general, con métricas muy equilibradas en cuanto a precisión, recall y F1 score. La matriz de confusión muestra que el modelo tiene un bajo número de falsos negativos (FN) y falsos positivos (FP), lo que indica que está haciendo predicciones correctas en su mayoría.

 En este caso, la matriz de confusión muestra 50 verdaderos negativos (TN), 13 falsos positivos (FP), 8 falsos negativos (FN) y 88 verdaderos positivos (TP). A partir de estos valores, podemos calcular métricas importantes como la precisión y la accuracy ,  el modelo predice la clase positiva, tiene un 86.76% de probabilidad de ser correcto.

Con accuracy, se pudo medir la proporción total de predicciones correctas en relación con todas las muestras, con un valor de 86.79%. Esta métrica muestra que el modelo ha acertado en un 86.79% de las veces, lo que refleja un buen rendimiento general. Sin embargo, el análisis de la matriz de confusión también revela que el modelo tiene un número significativo de falsos positivos (13) y falsos negativos (8), lo que puede sugerir áreas de mejora, dependiendo de las prioridades del caso de uso, como minimizar los falsos negativos o falsos positivos.
